### PR TITLE
cleanup ParseOptionAndSeparators

### PR DESCRIPTION
**Pull request by SimonCropp**

### CODE REVIEW

**Comment by SimonCropp** [comment]
@evangelink  i need some help on this one. did i misinterpret what the previous code was meant to do?

File: src/Platform/Microsoft.Testing.Platform/CommandLine/Parser.cs [hunk 1]
```diff
@@ -122,20 +122,17 @@ public static CommandLineParseResult Parse(string[] args, IEnvironment environme
...
-
-            currentOption = arg[..(RoslynString.IsNullOrEmpty(delimiterFound) ? arg.Length : arg.IndexOf(delimiterFound, StringComparison.InvariantCultureIgnoreCase))].TrimStart('-');
+            int delimiterIndex = arg.IndexOfAny([':', '=', ' ']);
 
-            if (!RoslynString.IsNullOrEmpty(delimiterFound))
+            if (delimiterIndex == -1)
             {
-                currentArg = arg[(arg.IndexOf(delimiterFound, StringComparison.InvariantCultureIgnoreCase) + 1)..];
+                currentArg = null;
+                currentOption = arg;
```
Comment (Evangelink) [inline src/Platform/Microsoft.Testing.Platform/CommandLine/Parser.cs]: This was set to `arg.Length`

**Review by Evangelink — COMMENTED** [review]
Otherwise no the logic seems good to me.

One of the failure is:

```
.NET Testing Platform v3.5.0-ci (UTC 2024/05/29) [win10-x64 - .NET 6.0.30]
Unknown option '----report-trx'
Unknown option '----report-trx-filename'
Unknown option '----results-directory'
Unknown option '----coverage'
Unknown option '----coverage-settings'
Unknown option '----coverage-output'

Usage dotnet exec Microsoft.Testing.Platform.UnitTests.dll [option providers] [extension option providers]

Execute a .NET Test Application.

Options:
  --diagnostic                             Enable the diagnostic logging. The default log level is 'Trace'. The file will be written in the output directory with the name log_[MMddHHssfff].diag
  --diagnostic-filelogger-synchronouswrite Force the built-in file logger to write the log synchronously. Useful for scenario where you don't want to lose any log (i.e. in case of crash). Note that this is slowing down the test execution.
  --diagnostic-output-directory            Output directory of the diagnostic logging, if not specified the file will be generated inside the default 'TestResults' directory.
  --diagnostic-output-fileprefix           Prefix for the log file name that will replace '[log]_.'
  --diagnostic-verbosity                   Define the level of the verbosity for the --diagnostic. The available values are 'Trace', 'Debug', 'Information', 'Warning', 'Error', and 'Critical'
  --exit-on-process-exit                   Exit the test process if dependent process exits. PID must be provided.
  --help                                   Show the command line help.
  --ignore-exit-code                       Do not report non successful exit value for specific exit codes (e.g. '--ignore-exit-code 8;9' ignore exit code 8 and 9 and will return 0 in these case)
  --info                                   Display .NET test application information.
  --list-tests                             List available tests.
  --minimum-expected-tests                 Specifies the minimum number of tests that are expected to run.
  --results-directory                      The directory where the test results are going to be placed. If the specified directory doesn't exist, it's created. The default is TestResults in the directory that contains the test application.

Extension options:
  --baseline-trx           The baseline TRX file
  --coverage               Collect the code coverage using dotnet-coverage tool
  --coverage-output        Output file
  --coverage-output-format Output file format. Supported values: 'coverage', 'xml' and 'cobertura'
  --coverage-settings      XML code coverage settings
  --crashdump              [net6.0+ only] Generate a dump file if the test process crashes
  --crashdump-filename     Specify the name of the dump file
  --crashdump-type         Specify the type of the dump. Valid values are 'Mini', 'Heap', 'Triage' or 'Full'. Default type is 'Full'. 
For more information visit https://learn.microsoft.com/dotnet/core/diagnostics/collect-dumps-crash#types-of-mini-dumps
  --report-trx             Enable generating TRX report
  --report-trx-filename    The name of the generated TRX report
  --treenode-filter        Use a tree filter to filter down the tests to execute
  --trx-to-compare         The TRX file to compare with the baseline
```

so I would say that the parsing is either not escaping correctly or the tests need to be updated.

I sadly won't have time to debug this today.

File: src/Platform/Microsoft.Testing.Platform/CommandLine/Parser.cs (hunk 1 above)
Comment (Evangelink) [reply src/Platform/Microsoft.Testing.Platform/CommandLine/Parser.cs]:
```suggestion
                currentOption = arg.Length;
```

**Review by Evangelink — APPROVED** [review]

File: src/Platform/Microsoft.Testing.Platform/CommandLine/Parser.cs [hunk 2]
```diff
@@ -122,20 +122,17 @@ public static CommandLineParseResult Parse(string[] args, IEnvironment environme
...
-            currentArg = null;
-            (int delimiterIndex, string delimiterFound) = new List<(int, string)>
-            {
-                (arg.IndexOf(":", StringComparison.InvariantCultureIgnoreCase), ":"),
-                (arg.IndexOf("=", StringComparison.InvariantCultureIgnoreCase), "="),
-                (arg.IndexOf(" ", StringComparison.InvariantCultureIgnoreCase), " "),
-            }.Where(x => x.Item1 != -1).OrderBy(x => x.Item1).FirstOrDefault();
-
-            currentOption = arg[..(RoslynString.IsNullOrEmpty(delimiterFound) ? arg.Length : arg.IndexOf(delimiterFound, StringComparison.InvariantCultureIgnoreCase))].TrimStart('-');
+            int delimiterIndex = arg.IndexOfAny([':', '=', ' ']);
```
Comment (Evangelink) [inline src/Platform/Microsoft.Testing.Platform/CommandLine/Parser.cs]:
```suggestion
            (currentOption, currentArg) = arg.IndexOfAny([':', '=', ' ']) switch
            {
                -1 => (arg, null),
                var delimiterIndex => (arg[..delimiterIndex], arg[(delimiterIndex + 1)..]),
            };

            currentOption = currentOption.TrimStart('-');
```

File: src/Platform/Microsoft.Testing.Platform/CommandLine/Parser.cs [hunk 3]
```diff
@@ -122,20 +122,17 @@ public static CommandLineParseResult Parse(string[] args, IEnvironment environme
...
+            if (delimiterIndex == -1)
             {
-                currentArg = arg[(arg.IndexOf(delimiterFound, StringComparison.InvariantCultureIgnoreCase) + 1)..];
+                currentArg = null;
+                currentOption = arg.Length;
+                return;
             }
+
+            currentOption = arg[..delimiterIndex].TrimStart('-');
+            currentArg = arg[(delimiterIndex + 1)..];
```
Comment (Evangelink) [inline src/Platform/Microsoft.Testing.Platform/CommandLine/Parser.cs]:
```suggestion
```

File: src/Platform/Microsoft.Testing.Platform/CommandLine/Parser.cs [hunk 4]
```diff
@@ -122,20 +122,17 @@ public static CommandLineParseResult Parse(string[] args, IEnvironment environme
...
-                (arg.IndexOf("=", StringComparison.InvariantCultureIgnoreCase), "="),
-                (arg.IndexOf(" ", StringComparison.InvariantCultureIgnoreCase), " "),
-            }.Where(x => x.Item1 != -1).OrderBy(x => x.Item1).FirstOrDefault();
-
-            currentOption = arg[..(RoslynString.IsNullOrEmpty(delimiterFound) ? arg.Length : arg.IndexOf(delimiterFound, StringComparison.InvariantCultureIgnoreCase))].TrimStart('-');
+            int delimiterIndex = arg.IndexOfAny([':', '=', ' ']);
 
-            if (!RoslynString.IsNullOrEmpty(delimiterFound))
+            if (delimiterIndex == -1)
             {
```
Comment (Evangelink) [inline src/Platform/Microsoft.Testing.Platform/CommandLine/Parser.cs]:
```suggestion
```

File: src/Platform/Microsoft.Testing.Platform/CommandLine/Parser.cs [hunk 5]
```diff
@@ -122,20 +122,14 @@ public static CommandLineParseResult Parse(string[] args, IEnvironment environme
...
-                (arg.IndexOf(":", StringComparison.InvariantCultureIgnoreCase), ":"),
-                (arg.IndexOf("=", StringComparison.InvariantCultureIgnoreCase), "="),
-                (arg.IndexOf(" ", StringComparison.InvariantCultureIgnoreCase), " "),
-            }.Where(x => x.Item1 != -1).OrderBy(x => x.Item1).FirstOrDefault();
+                -1 => (arg, null),
+                var delimiterIndex => (arg[..delimiterIndex], arg[(delimiterIndex + 1)..]),
+            };
 
-            currentOption = arg[..(RoslynString.IsNullOrEmpty(delimiterFound) ? arg.Length : arg.IndexOf(delimiterFound, StringComparison.InvariantCultureIgnoreCase))].TrimStart('-');
+            currentOption = currentOption.TrimStart('-');
```
Comment (Evangelink) [inline src/Platform/Microsoft.Testing.Platform/CommandLine/Parser.cs]:
```suggestion
```

**Review by Evangelink — APPROVED** [review]

**Pull request merged**